File: Sources/Wrap/GitWrap.swift
import Foundation
import IOUtils

public class GitWrap: VcsWrap {
  public private(set) var depth: Int
  public private(set) var pushURL: String?
  public private(set) var cloneRecursive: Bool

  internal init(
    wrapHash: String,
    directory: String?,
    patchURL: String?,
    patchFallbackURL: String?,
    patchFilename: String?,
    patchHash: String?,
    patchDirectory: String?,
    diffFiles: [String],
    url: String?,
    revision: String?,
    depth: Int,
    pushURL: String?,
    cloneRecursive: Bool
  ) {
    self.depth = depth
    self.pushURL = pushURL
    self.cloneRecursive = cloneRecursive
    super.init(
      wrapHash: wrapHash,
      directory: directory,
      patchURL: patchURL,
      patchFallbackURL: patchFallbackURL,
      patchFilename: patchFilename,
      patchHash: patchHash,
      patchDirectory: patchDirectory,
      diffFiles: diffFiles,
      url: url,
      revision: revision
    )
  }

  public override func setupDirectory(path: String, packagefilesPath: String) throws {
    try self.assertRequired("git")
    guard let urlAsString = self.url else { throw WrapError.genericError("Expected URL to clone") }
    if let url = URL(string: urlAsString) {
      guard let rev = self.revision else { throw WrapError.genericError("Missing revision") }
      // Do something like https://github.com/mesonbuild/meson/blob/3e7c08f358e9bd91808c8ff3b76c11aedeb82f85/mesonbuild/wrap/wrap.py#L549
      let targetDirectory =
        self.directory ?? url.lastPathComponent.replacingOccurrences(of: ".git", with: "")
      let fullPath = "\(path)\(Path.separator)\(targetDirectory)"
      let isShallow = self.depth != 0 && self.depth != Int.max
      let depthOptions = isShallow ? ["--depth", self.depth.description] : []
      if isShallow && self.isValidCommitId(rev) {
        try self.executeCommand([
          "git", "-c", "init.defaultBranch=mesonlsp-dummy-branch", "init", fullPath,
        ])
        try self.executeCommand(["git", "-C", fullPath, "remote", "add", "origin", urlAsString])
        try self.executeCommand(["git", "-C", fullPath, "fetch"] + depthOptions + ["origin", rev])
        try self.executeCommand([
          "git", "-C", fullPath, "-c", "advice.detachedHead=false", "checkout", rev, "--",
        ])
      } else {
        if !isShallow {
          try self.executeCommand(["git", "clone", urlAsString, fullPath])
          if rev.lowercased() != "head" {
            do {
              try self.executeCommand([
                "git", "-C", fullPath, "-c", "advice.detachedHead=false", "checkout", rev, "--",
              ])
            } catch {
              try self.executeCommand(["git", "-C", fullPath, "fetch", urlAsString, rev])
              try self.executeCommand([
                "git", "-C", fullPath, "-c", "advice.detachedHead=false", "checkout", rev, "--",
              ])
            }
          }
        } else {
          var args = ["-c", "advice.detachedHead=false", "clone"] + depthOptions
          if rev.lowercased() != "head" { args += ["--branch", rev] }
          args += [urlAsString, fullPath]
          try self.executeCommand(["git"] + args)
        }
      }
      if self.cloneRecursive {
        try self.executeCommand(
          ["git", "-C", fullPath, "submodule", "update", "--init", "--checkout", "--recursive"]
            + depthOptions
        )
      }
      if let pl = self.pushURL {
        try self.executeCommand([
          "git", "-C", fullPath, "remote", "set-url", "--push", "origin", pl,
        ])
      }
      try self.postSetup(path: fullPath, packagesfilesPath: packagefilesPath)
    } else {
      throw WrapError.genericError("Malformed URL: \(String(describing: self.url))")
    }
  }

  private func isValidCommitId(_ id: String) -> Bool {
    if id.count != 40 && id.count != 64 { return false }
    return id.filter { $0.isHexDigit }.count == id.count
  }
}

File: src/libwrap/gitwrap.cpp
#include "ini.hpp"
#include "log.hpp"
#include "polyfill.hpp"
#include "utils.hpp"
#include "wrap.hpp"

#include <algorithm>
#include <cctype>
#include <charconv>
#include <filesystem>
#include <fstream>
#include <string>
#include <system_error>
#include <vector>

const static Logger LOG("wrap::GitWrap"); // NOLINT

GitWrap::GitWrap(ast::ini::Section *node) : VcsWrap(node) {
  if (auto optPushUrl = node->findStringValue("push-url")) {
    this->pushUrl = optPushUrl;
  }
  if (auto depthString = node->findStringValue("depth")) {
    int number;
    auto res = std::from_chars(
        depthString->data(), depthString->data() + depthString->size(), number);
    if (res.ec == std::errc{}) {
      this->depth = number;
    }
  }
  if (auto val = node->findStringValue("clone-recursive")) {
    this->cloneRecursive = val == "true";
  }
}

static bool isValidCommitId(std::string rev) {
  if (rev.size() != 40 && rev.size() != 64) {
    return false;
  }
  return std::ranges::all_of(rev, [](char chr) { return std::isxdigit(chr); });
}

static bool isHead(std::string rev) {
  return rev.size() == 4 &&
         (std::tolower(rev[0]) == 'h' && std::tolower(rev[1]) == 'e' &&
          std::tolower(rev[2]) == 'a' && std::tolower(rev[3]) == 'd');
}

bool GitWrap::setupDirectory(const std::filesystem::path &path,
                             const std::filesystem::path &packageFilesPath) {
  auto url = this->url;
  if (url.empty()) {
    LOG.warn("URL is empty");
    this->errors.emplace_back("Missing url");
    return false;
  }
  if (this->revision.empty()) {
    LOG.warn("Revision is empty");
    this->errors.emplace_back("Missing revision");
    return false;
  }
  if (!this->directory.has_value()) {
    LOG.warn("Directory is empty");
    this->errors.emplace_back("Missing directory");
    return false;
  }
  auto rev = this->revision;
  auto targetDirectory = this->directory.value();
  std::filesystem::path const fullPath =
      std::format("{}/{}", path.generic_string(), targetDirectory);
  auto isShallow = this->depth != 0;
  auto depthOptions =
      isShallow
          ? std::vector<std::string>{"--depth", std::to_string(this->depth)}
          : std::vector<std::string>{};
  if (isShallow && isValidCommitId(rev)) {
    auto result = launchProcess(
        "git", std::vector<std::string>{
                   "-c", "init.defaultBranch=mesonlsp-dummy-branch", "init",
                   fullPath.generic_string()});
    if (!result) {
      this->errors.emplace_back("Failed to do git init");
      return false;
    }
    result = launchProcess(
        "git", std::vector<std::string>{"-C", fullPath.generic_string(),
                                        "remote", "add", "origin", this->url});
    if (!result) {
      this->errors.emplace_back("Failed to do remote add origin");
      return false;
    }
    auto fetchOptions =
        std::vector<std::string>{"-C", fullPath.generic_string(), "fetch"};
    fetchOptions.insert(fetchOptions.end(), depthOptions.begin(),
                        depthOptions.end());
    fetchOptions.emplace_back("origin");
    fetchOptions.push_back(rev);
    result = launchProcess("git", fetchOptions);
    if (!result) {
      this->errors.emplace_back("Failed to do git fetch origin");
      return false;
    }
    result = launchProcess(
        "git", std::vector<std::string>{"-C", fullPath.generic_string(), "-c",
                                        "advice.detachedHead=false", "checkout",
                                        rev, "--"});
    if (!result) {
      this->errors.emplace_back("Failed to do git checkout");
      return false;
    }
  } else {
    if (!isShallow) {
      auto result = launchProcess(
          "git", std::vector<std::string>{"clone", this->url,
                                          fullPath.generic_string()});
      if (!result) {
        this->errors.emplace_back("Failed to do git clone");
        return false;
      }
      if (!isHead(rev)) {
        result = launchProcess(
            "git", std::vector<std::string>{"-C", fullPath.generic_string(),
                                            "-c", "advice.detachedHead=false",
                                            "checkout", rev, "--"});
        if (!result) {
          result = launchProcess(
              "git", std::vector<std::string>{"-C", fullPath.generic_string(),
                                              "fetch", this->url, rev});
          if (!result) {
            this->errors.emplace_back("Failed to do git checkout/fetch");
            return false;
          }
          result = launchProcess(
              "git", std::vector<std::string>{"-C", fullPath.generic_string(),
                                              "-c", "advice.detachedHead=false",
                                              "checkout", rev, "--"});
          if (!result) {
            this->errors.emplace_back("Failed to do git checkout/checkout");
            return false;
          }
        }
      }
    } else {
      std::vector<std::string> args{"-c", "advice.detachedHead=false", "clone"};
      args.insert(args.end(), depthOptions.begin(), depthOptions.end());
      if (!isHead(rev)) {
        args.emplace_back("--branch");
        args.push_back(rev);
      }
      args.push_back(this->url);
      args.push_back(fullPath.generic_string());
      auto result = launchProcess("git", args);
      if (!result) {
        this->errors.emplace_back("Failed to do git clone with a branch");
        return false;
      }
    }
  }
  if (this->cloneRecursive) {
    auto cloneOptions =
        std::vector<std::string>{"-C",         fullPath.generic_string(),
                                 "submodule",  "update",
                                 "--init",     "--checkout",
                                 "--recursive"};
    cloneOptions.insert(cloneOptions.end(), depthOptions.begin(),
                        depthOptions.end());
    auto result = launchProcess("git", cloneOptions);
    if (!result) {
      this->errors.emplace_back("Failed to update git submodules");
      return false;
    }
  }
  if (auto optPushUrl = this->pushUrl) {
    auto result = launchProcess(
        "git", std::vector<std::string>{"-C", fullPath.generic_string(),
                                        "remote", "set-url", "--push", "origin",
                                        optPushUrl.value()});
    if (!result) {
      this->errors.emplace_back("Failed to set push url");
      return false;
    }
  }
  if (!this->postSetup(fullPath, packageFilesPath)) {
    return false;
  }
  if (isValidCommitId(rev)) {
    return true;
  }
  auto result = launchProcess(
      "git", std::vector<std::string>{"-C", fullPath.generic_string(), "pull",
                                      "origin"});
  if (result) {
    auto pullableFile = fullPath / ".git_pullable";
    std::ofstream{pullableFile}.put('\n');
  }
  // If it fails, it was a tag, not a branch
  return true;
}
